examples/tutorials/device_asr.py

```diff
@@ -0,0 +1,166 @@
+    """Cache the previous chunk and combine it with the new chunk
+
+    Args:
+        chunk (torch.Tensor): Initial chunk
+        right_context_size (int): The size of right context.
+    """
+
+    def __init__(self, chunk: torch.Tensor, right_context_size: int):
+        self.chunk = chunk
+        self.right_context_size = right_context_size
+
+    def __call__(self, chunk: torch.Tensor):
+        right_context = chunk[: self.right_context_size, :]
+        chunk_with_context = torch.cat((self.chunk, right_context))
+        self.chunk = chunk
+        return chunk_with_context
+
+
+######################################################################
+# 4. Configuring mirophone stream
+# -------------------------------
+#
+# Next, we configure the input audio stream using
+# :py:func:`~torchaudio.prototype.io.Streamer`.
+# For the basic usage of Streamer API, please refer to
+# `Media Stream API tutorial <./streaming_api_tutorial.html>`__.
+#
+# To connect to media device, we provide ``format`` and ``option``
+# parameters to the constructor.
+# The exact arguments to be passed depend on the system configuration.
+# Please refer to https://ffmpeg.org/ffmpeg-devices.html for the detail.
+#
+# For macOS, the supported ``format`` value is ``"avfoundation"``.
+# The following command lists the available devices.
+#
+# .. code::
+#
+#    $ ffmpeg -f avfoundation -list_devices true -i ""
+#    [AVFoundation indev @ 0x126e049d0] AVFoundation video devices:
+#    [AVFoundation indev @ 0x126e049d0] [0] FaceTime HD Camera
+#    [AVFoundation indev @ 0x126e049d0] [1] Capture screen 0
+#    [AVFoundation indev @ 0x126e049d0] AVFoundation audio devices:
+#    [AVFoundation indev @ 0x126e049d0] [0] ZoomAudioDevice
+#    [AVFoundation indev @ 0x126e049d0] [1] MacBook Pro Microphone
+#
+# We use `MacBook Pro Microphone` as audio device (index 1).
+#
+
+device, src = "avfoundation", ":1"
+streamer = Streamer(src, format=device)
+streamer.add_basic_audio_stream(frames_per_chunk=frames_per_chunk, sample_rate=sample_rate)
+
+print(streamer.get_src_stream_info(0))
+print(streamer.get_out_stream_info(0))
+
+######################################################################
+# .. code::
+#
+#    >>> SourceAudioStream(media_type='audio', codec='pcm_f32le', codec_long_name='PCM 32-bit floating point little-endian', format='flt', bit_rate=1536000, sample_rate=48000.0, num_channels=1)
+#    >>> OutputStream(source_index=0, filter_description='aresample=16000,aformat=sample_fmts=fltp')
+#
+
+######################################################################
+# 5. Run stream inference
+# -----------------------
+#
+# Finally, we run the recognition.
+#
+# First, we initialize the stream iterator, context cacher, and
+# state and hypothesis that are used by decoder to carry over the
+# decoding state between inference calls.
+#
+
+stream_iterator = streamer.stream(retry=1)
+cacher = ContextCacher(next(stream_iterator)[0], right_context_size)
```

Comment:
one observation: we read two chunks before running inference on the first chunk, but we actually only need the first chunk plus the right context. the former adds a delay (constant offset) of `chunk length - right context length` to the transcriptions, which in this case is 120ms and would be amplified as chunk length increases

ideally, we'd initialize `ContextCacher` with the first `right context length` amount of data in the stream, and then proceed with processing a chunk on each call as is currently done

Reply:
Very good point. In fact, I was a bit confused about the fact it is called "right context", as opposed to "left context". "left" would infer that it is a carry over from the past. This terminology had me think in this way.

I think we can initialize the cacher with zero Tensors of `right context length`, then start feeding, append and cache the latest. Let me try this on the existing tutorial.